lib/Sema/TypeCheckMacros.cpp

```diff
@@ -919,53 +924,76 @@ swift::expandMacroExpr(MacroExpansionExpr *mee) {
   case MacroDefinition::Kind::External: {
     // Retrieve the external definition of the macro.
     auto external = macroDef.getExternalMacro();
-    ExternalMacroDefinitionRequest request{
-      &ctx, external.moduleName, external.macroTypeName
-    };
+    ExternalMacroDefinitionRequest request{&ctx, external.moduleName,
+                                           external.macroTypeName};
     auto externalDef = evaluateOrDefault(ctx.evaluator, request, None);
     if (!externalDef) {
-      ctx.Diags.diagnose(
-          mee->getLoc(), diag::external_macro_not_found,
-          external.moduleName.str(),
-          external.macroTypeName.str(),
-          macro->getName()
-      );
+      ctx.Diags.diagnose(loc, diag::external_macro_not_found,
+                         external.moduleName.str(),
+                         external.macroTypeName.str(), macro->getName());
       macro->diagnose(diag::decl_declared_here, macro->getName());
-      return None;
+      return nullptr;
+    }
+
+    // Currently only expression macros are enabled by default. Declaration
+    // macros need the `FreestandingMacros` feature flag, and code item macros
+    // need both `FreestandingMacros` and `CodeItemMacros`.
+    if (!macroRoles.contains(MacroRole::Expression)) {
+      if (!macroRoles.contains(MacroRole::Declaration) &&
+          !ctx.LangOpts.hasFeature(Feature::CodeItemMacros)) {
+        ctx.Diags.diagnose(loc, diag::macro_experimental, "code item",
+                           "CodeItemMacros");
+        return nullptr;
+      }
```

Comment:
I don’t think the implementation matches the comment here. Declaration macros don’t emit a diagnostic. 

AFAIK declaration macros are enabled by default now, so I think you should be able to simplify this to

```suggestion
    if (macroRoles.contains(MacroRole::CodeItem) &&
        !ctx.LangOpts.hasFeature(Feature::CodeItemMacros)) {
      ctx.Diags.diagnose(loc, diag::macro_experimental, "code item",
                         "CodeItemMacros");
      return nullptr;
    }
```


---

And yes, I just saw that you just copy-pasted this from the old implementation.

Reply:
Yeah I just moved this code. But let's do it in this PR. updated.